### ISSUE TITLE
Deprecation of next_transparent parameter of Model#start_operation method

### ISSUE THREAD

**Issue by esterkimx** [SketchUp/api-issue-tracker#6]
The `next_transparent` parameter for `Model#start_operation` needed.  

The `Model#start_operation` method provides some control for plugin developers over the SketchUp undo stack. It has four parameters:

- _op_name (String)_ 
- _disable_ui (Boolean) (defaults to: false)_
- _next_transparent (Boolean) (defaults to: false)_ 
- _transparent (Boolean) (defaults to: false)_

The `next_transparent` parameter is currently marked as ‘Deprecated’ in the API documentation and even has a ‘Use with extreme caution’ label.  We have found ourselves in situations where we cannot do some operations without the `next_transparent` argument set to true.
 
For example, when we want to load a component from a file and place it into the model. We do it when a user presses a button on the toolbar - so we want that the loading and the placement appear as a single operation in the undo stack. 

The following code example loads and places the component, but it commits two operations into the undo stack:

```ruby
model = Sketchup.active_model
component_def = model.definitions.load(path_to_component_file)
model.place_component(component_def)
```

We can try and make the `#place_component` invocation transparent:

```ruby
component_def = model.definitions.load(path_to_component_file)
model.start_operation "Loading and placing component",
                                    transparent=true
model.place_component(component_def)
model.commit_operation
```

But this doesn’t work, because operations in SketchUp cannot be nested. 

The following solution resolves the problem:

```ruby
model.start_operation "Transparent component placement",
                                    disable_ui=false,
                                    next_transparent=true,
                                    transparent=false
component_def = model.definitions.load(path_to_component_file)
model.commit_operation
model.place_component(component_def)
```

Since both `Model#place_component` and `DefinitionList#load` methods commit operations into the stack and operations cannot be nested, it’s impossible to unite their invocation into one operation without the next_transparent parameter. 

Please make sure not to remove this parameter in future versions of the Sketchup API.

**Comment by thomthom** [comment]
Yes, this is one of the rare cases where this parameter is needed to account for a flaw in the API elsewhere.

We don't intend on removing it from the Ruby API, but we've added stern warnings because we've encountered a number of problems as a result of people using this parameter without fully understanding what it does.

**Comment by frandzel3d** [comment]
We posted this following your advice, just to make sure it's reported, known, and not changing, because it can seriously affect our extension. 
Thanks Thomthom.

**Comment by DanRathbun** [comment]
Here is a test script defining 3 methods that show the weird behavior described in the 1st post.

[load_and_place_comp_one_oper.zip](https://github.com/SketchUp/api-issue-tracker/files/1514907/load_and_place_comp_one_oper.zip)

-----

**Just thinking out loud** ...

Changing these two API methods would break code "in the wild". 

What about creating a new block form method that allow the grouping of methods that cause a commit ?
```ruby
transparently do
  component_def = model.definitions.load(path_to_component_file)  
  model.place_component(component_def)
end
```
Thinking that when statements are executed within this `transparently()` method, that undo commits are suppressed.

Then the above method call could be used within a normal API operation syntax.

**Comment by thomthom** [comment]
The problem is that `model.place_component(component_def)` cause a tool to activate - and the new operation isn't committed until that tool completes.  In between that other operation could happen - or even the tool cancels.

`definition.load` and `place_component` is a recurring combo which I think is unique. Maybe we need to create a special subclass of the Place Component tool which takes care of handling the loading of definitions and placement for the APIs.

**Comment by DanRathbun** [comment]
But where would the new method be implemented ? model or definitions object ?

I like the idea, and it should return array :
```ruby
[ instance, instance, ... ]
```
If the tool is cancelled, it should undo the component load.

**Comment by thomthom** [comment]
Maybe an additional optional parameter to `place_component` - since it would behave much the same, just take care of loading the definition as well.

**Comment by DanRathbun** [comment]
That sounds like complicating a simple thing.

How do you choose between local load and load_url ?
And how does a coder leverage a load handler for the latter case ?
(A load handler reference would definitely need to be a third argument.)

I could see a conditional upon the 1st arg, that if it's a definition reference it acts normal, but if it's a String pathname to an SKP file, it does a load (local).

Is it easy for you to tell a local pathname from a remote url ?
```ruby
PROTOCOLS ||= /\A(http|ftp|\/)/i  # starts with /, http, ftp, ...
```
?

Back to the idea of changing `Model#place_component()` ...

You can just ducktype the method name if it gets a new method for newer SU versions.

But changing an old method will cause inability for duck-typing. 
And we **cannot** ducktype the argument number because it's both variable and defined in **C**, viz:
```ruby
model.method(:place_component).arity
#=> -1
model.method(:place_component).parameters
#=> [[:rest]]
```

-----

I think it'd be smarter to add a new method perhaps named:
* **`Model#load_and_place_component()`**
* **`Model#place_after_load_component()`**

?

**Comment by DanRathbun** [comment]
> definition.load and place_component is a recurring combo which I think is unique.

And just a note ... that `definition.load()` was causing this same error all by itself until v14 ...

https://forums.sketchup.com/t/definitionlist-load-breaking-undo-stack/39573/8

**Comment by DanRathbun** [comment]
Also as reference, an example "MoveTool" clone that shows the "hoops" a coder must "jump through" in order to wrap `place_component()` into a single command operation.

I wrote this back in June of 2016:
https://forums.sketchup.com/t/code-undoable-place-component-movetool-clone/27193

**Comment by DanRathbun** [comment]
Please log new method enhancement that you yourself proposed ...

> @thomthom definition.load and place_component is a recurring combo which I think is unique. Maybe we need to create a special subclass of the Place Component tool which takes care of handling the loading of definitions and placement for the APIs.

I noted that changing the current method will cause issues ...

> @DanRathbun But changing an old method will cause inability for duck-typing.
And we cannot ducktype the argument number because it's both variable and defined in C

And, I proposed 2 possible names above ...

> `Model#load_and_place_component( path_or_url, repeat, handler )`
> `Model#place_after_load_component( path_or_url, repeat, handler )`

**Comment by thomthom** [comment]
Logged as SU-40235

**Comment by DanRathbun** [comment]
Thinking that a handler object should accept the definition, new instance, it's index in the current repeat loop, and it's num in the set of it's definition's instances (ie, the size of the set being `definition.instances.size`.)

```ruby
class MyPlaceComponentHandler

  def initialize(material)
    @material = material
  end

  def onPlaceComponent(definition, instance, index, num)
    if num > 1
      instance.name= instance.definition.name << ' - ' << num.to_s
    else
      instance.name= instance.definition.name
    end
    instance.material= @material
  end

  def onCancel(definition, reason, index, num)
    # react according to the reasons given for Tool#onCancel callback
  end

  def onCompletion(definition, added, total)
    # runs after onCancel or immediately after a single place (ie, repeat was false)
  end

end
```

A handler class object is likely to be more versatile than a block form.
